File: src/main/scala/com/bryghts/kissjson/parser2/tools/fsb/FastStringBuffer.java
package com.bryghts.kissjson.parser2.tools.fsb;

import java.util.Arrays;

public class FastStringBuffer
{
	private int l = 0;
	private char buffer[];

	private int limit;// =

    public FastStringBuffer (int initialCapacity)
    {
        buffer = new char[(initialCapacity < 10)? 10 : initialCapacity];
        limit = ((initialCapacity < 10)? 10 : initialCapacity) - 2;
    }

	public void clear() {l = 0;}

    public void revert() {
        l--;
    }
	public void append (char c)
	{

		if(l >= limit)
		{
			int s = buffer.length << 1;
			buffer = Arrays.copyOf(buffer, s);
			limit = s - 2;
		}

		buffer[l] = c;
		l ++;

	}

	@Override public String toString(){return new String(buffer, 0, l);}
}

File: src/main/scala/com/bryghts/kissjson/parser2/tools/fsb/FastStringBuffer.scala
//package com.bryghts.kissjson.parser2.tools.fsb
//
//import com.bryghts.kissjson.parser2.tools.uchar.UChar
//import java.util.Arrays
//
//class FastStringBuffer (initialCapacity: Int)
//{
//	private var l = 0
//	private var buffer: Array[Char] = new Array[Char](if(initialCapacity < 10) 10 else initialCapacity)
//	private var limit = (if(initialCapacity < 10) 10 else initialCapacity) - 2
//
//	def clear() = l = 0
//
//	def += (c: UChar):Unit =
//	{
//
//		if(l >= limit)
//		{
//			val s = buffer.length << 1
//			buffer = Arrays.copyOf(buffer, s)
//			limit = s - 2
//		}
//
//		buffer(l) = c
//		l += 1
//
//	}
//
//	@inline def append (c: UChar):Unit =
//	{
//
//		if(l >= limit)
//		{
//			val s = buffer.length << 1
//			buffer = Arrays.copyOf(buffer, s)
//			limit = s - 2
//		}
//
//		buffer(l) = c
//		l += 1
//
//	}
//
//	private def ensureCapacity()
//	{
//		if(l >= limit)
//		{
//			val s = buffer.length << 1
//			buffer = Arrays.copyOf(buffer, s)
//			limit = s - 2
//		}
//	}
//
//	override def toString(): String = new String(buffer, 0, l)
//}
//
//object FastStringBuffer {
//	def apply(initialCapacity: Int) = new FastStringBuffer(initialCapacity)
//}